Disable Babel polyfill injection in dev

renchap opened the pull request:
In https://github.com/mastodon/mastodon/pull/27333 I added `useBuiltIns: "usage"`, which causes Babel to inject polyfills into those files. The polyfills are injected using ESM `import` symtax, which disables the handling of `export.` (CommonJS exports).

This causes the build in dev to fail, because in dev Babel does not convert the exports to CJS (not sure why).

I tried various things but I am unsure of why this happens. A simple fix is to no longer have Babel inject polyfills in dev, which was the previous behaviour.

We have plans to rewrite the emoji-related code in the future, which will remove the need for `preval`, and switching to Vite should also solve this.

Fixes #27686

renchap commented:
Note: I also replaced the custom `padLeft` with `padStart`.

jamesadney commented:
I get a new error running the server in development on this branch:

```
Failed to compile.

./app/javascript/mastodon/features/emoji/emoji_compressed.js
Module build failed (from ./node_modules/babel-loader/lib/index.js):
/tmp/mastodon/app/javascript/mastodon/features/emoji/unicode_to_names.js:3
export const unicodeToFilename = (str) => {
^^^^^^

SyntaxError: Unexpected token 'export'
    at internalCompileFunction (node:internal/vm:73:18)
    at wrapSafe (node:internal/modules/cjs/loader:1153:20)
    at Module._compile (node:internal/modules/cjs/loader:1205:27)
    at Module._extensions..js (node:internal/modules/cjs/loader:1295:10)
    at Module.load (node:internal/modules/cjs/loader:1091:32)
    at Module._load (node:internal/modules/cjs/loader:938:12)
    at Module.require (node:internal/modules/cjs/loader:1115:19)
    at require (node:internal/modules/helpers:130:18)
    at Object.<anonymous> (/tmp/mastodon/app/javascript/mastodon/features/emoji/unicode_to_names_cjs.js:4:53)
    at Module._compile (node:internal/modules/cjs/loader:1241:14)
    at Module._extensions..js (node:internal/modules/cjs/loader:1295:10)
    at Module.load (node:internal/modules/cjs/loader:1091:32)
    at Module._load (node:internal/modules/cjs/loader:938:12)
    at Module.require (node:internal/modules/cjs/loader:1115:19)
    at require (node:internal/modules/helpers:130:18)
    at Object.<anonymous> (/tmp/mastodon/app/javascript/mastodon/features/emoji/emoji_compressed.js:21:5)
    at Module._compile (node:internal/modules/cjs/loader:1241:14)
    at requireFromString (/tmp/mastodon/node_modules/require-from-string/index.js:28:4)
    at requireFromString (/tmp/mastodon/node_modules/babel-plugin-preval/dist/helpers.js:35:47)
    at getReplacement (/tmp/mastodon/node_modules/babel-plugin-preval/dist/helpers.js:58:18)
    at PluginPass.Program (/tmp/mastodon/node_modules/babel-plugin-preval/dist/index.js:52:57)
    at newFn (/tmp/mastodon/node_modules/@babel/traverse/lib/visitors.js:159:14)
    at NodePath._call (/tmp/mastodon/node_modules/@babel/traverse/lib/path/context.js:46:20)
    at NodePath.call (/tmp/mastodon/node_modules/@babel/traverse/lib/path/context.js:36:17)
    at NodePath.visit (/tmp/mastodon/node_modules/@babel/traverse/lib/path/context.js:82:31)
    at TraversalContext.visitQueue (/tmp/mastodon/node_modules/@babel/traverse/lib/context.js:86:16)
    at TraversalContext.visitSingle (/tmp/mastodon/node_modules/@babel/traverse/lib/context.js:65:19)
    at TraversalContext.visit (/tmp/mastodon/node_modules/@babel/traverse/lib/context.js:109:19)
    at traverseNode (/tmp/mastodon/node_modules/@babel/traverse/lib/traverse-node.js:22:17)
    at traverse (/tmp/mastodon/node_modules/@babel/traverse/lib/index.js:52:34)
    at transformFile (/tmp/mastodon/node_modules/@babel/core/lib/transformation/index.js:82:31)
    at transformFile.next (<anonymous>)
    at run (/tmp/mastodon/node_modules/@babel/core/lib/transformation/index.js:24:12)
    at run.next (<anonymous>)
    at transform (/tmp/mastodon/node_modules/@babel/core/lib/transform.js:22:33)
    at transform.next (<anonymous>)
    at step (/tmp/mastodon/node_modules/gensync/index.js:261:32)
    at /tmp/mastodon/node_modules/gensync/index.js:273:13
    at async.call.result.err.err (/tmp/mastodon/node_modules/gensync/index.js:223:11)
    at /tmp/mastodon/node_modules/gensync/index.js:189:28
    at /tmp/mastodon/node_modules/@babel/core/lib/gensync-utils/async.js:68:7
    at /tmp/mastodon/node_modules/gensync/index.js:113:33
    at step (/tmp/mastodon/node_modules/gensync/index.js:287:14)
    at /tmp/mastodon/node_modules/gensync/index.js:273:13
    at async.call.result.err.err (/tmp/mastodon/node_modules/gensync/index.js:223:11)
```

renchap committed:
Disable Babel polyfills in dev

In #27333 I added useBuiltIns: "usage", which causes Babel to inject polyfills into those files. The polyfills are injected using ESM import symtax, which disables the handling of export. (CommonJS exports).

This causes the build in dev to fail, because in dev Babel does not convert the exports to CJS (not sure why).

I tried various things but I am unsure of why this happens. A simple fix is to no longer have Babel inject polyfills in dev, which was the previous behaviour.

zunda commented:
6311b6565ef419d2438b280c6be1b20eef9d2147 seems to work for me on Ubuntu 22.04.3, ruby 3.2.2 (2023-03-30 revision e51014f9c0) [x86_64-linux], node v20.8.0, Firefox 118.0.2 by:

```
$ cat ../mastodon.env.dev 
export RAILS_ENV=development
export NODE_ENV=development
$ . ../mastodon.env.dev
$ bundle exec rails assets:clean
$ bundle exec rails assets:precompile
$ foreman start
```

and super-reloading the browser at http://localhost:3000/home
